integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/columnLineage/InputFieldsCollector.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.spark.sql.catalyst.catalog.HiveTableRelation;
+import org.apache.spark.sql.catalyst.expressions.AttributeReference;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import org.apache.spark.sql.catalyst.plans.logical.UnaryNode;
+import org.apache.spark.sql.execution.LogicalRDD;
+import org.apache.spark.sql.execution.datasources.LogicalRelation;
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation;
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation;
+
+/** Traverses LogicalPlan and collect input fields with the corresponding ExprId. */
+@Slf4j
+class InputFieldsCollector {
+
+  private final LogicalPlan plan;
+  private final OpenLineageContext context;
+
+  InputFieldsCollector(LogicalPlan plan, OpenLineageContext context) {
+    this.plan = plan;
+    this.context = context;
+  }
+
+  void collect(ColumnLevelLineageBuilder builder) {
+    collect(plan, builder);
+  }
+
+  void collect(LogicalPlan node, ColumnLevelLineageBuilder builder) {
+    discoverInputsFromNode(node, builder);
+
+    // hacky way to replace `node instanceof UnaryNode` which fails for Spark 3.2.1
+    // because of java.lang.IncompatibleClassChangeError: UnaryNode, but class was expected
+    // probably related to single code base for different Spark versions
+    if ((node.getClass()).isAssignableFrom(UnaryNode.class)) {
+      collect(((UnaryNode) node).child(), builder);
+    } else if (node.children() != null) {
+      ScalaConversionUtils.<LogicalPlan>fromSeq(node.children()).stream()
+          .forEach(child -> collect(child, builder));
+    }
+  }
+
+  private void discoverInputsFromNode(LogicalPlan node, ColumnLevelLineageBuilder builder) {
+    extractDatasetIdentifier(node).stream()
+        .forEach(
+            di ->
+                ScalaConversionUtils.fromSeq(node.output()).stream()
+                    .filter(attr -> attr instanceof AttributeReference)
+                    .map(attr -> (AttributeReference) attr)
+                    .forEach(attr -> builder.addInput(attr.exprId(), di, attr.name())));
+  }
+
+  private List<DatasetIdentifier> extractDatasetIdentifier(LogicalPlan node) {
+    if (node instanceof DataSourceV2Relation) {
+      return extractDatasetIdentifier((DataSourceV2Relation) node);
+    } else if (node instanceof DataSourceV2ScanRelation) {
+      return extractDatasetIdentifier(((DataSourceV2ScanRelation) node).relation());
+    } else if (node instanceof HiveTableRelation) {
+      return extractDatasetIdentifier(((HiveTableRelation) node).tableMeta());
+    } else if (node instanceof LogicalRelation
+        && ((LogicalRelation) node).catalogTable().isDefined()) {
+      return extractDatasetIdentifier(((LogicalRelation) node).catalogTable().get());
+    } else if (node instanceof LogicalRDD) {
+      return extractDatasetIdentifier((LogicalRDD) node);
+    }
+    return Collections.emptyList();
```

Comment:
This is kind of tricky because when traversing the plan, there can be many plan nodes that we're just not interested in. Even logging all of them does not make much sense. 
We can go with `log.warn` for other `LeafNode` objects.